### ISSUE TITLE
Intercepting MPI calls made in external library

### ISSUE THREAD

**Issue by vineetsoni** [bsc-performance-tools/extrae#70]
**Note:** This problem is related to the usage of Extrae with our solver.

# Question
Is it possible to intercept the MPI calls made via an external library? If so, could you please explain how?

# MPI usage
**Normal case:** I have an external MPI library plugin (say e.g. `mpi-wrapper.so`) that I link to the solver (say e.g. `EXE`). At the runtime, the MPI call hierarchy is:  `EXE` -> `mpi-wrapper.so` -> `libmpi.so`. In this case, there is no MPI information recorded in the trace when I use Extrae.

**Test case:** If I change the MPI call hierarchy as `EXE` -> `libmpi.so`, MPI calls are intercepted by Extrae.

# Attempts to trace MPI calls
1. `LD_PRELOAD` => NOK
2. Linking `libmpitrace.so` to `EXE` => NOK
3. Linking `libmpitrace.so` to `EXE` and to `mpi-wrapper.so` => NOK

Is there any way to generate MPI information in the trace using Extrae for the _normal case_ mentioned above? Do you have any suggestions for this problem?

**Comment by emercadal** [comment]
How does the intermediate library call MPI? Does it use the profiling interface (PMPI)? Does it create function pointers to MPI routines and then call them using those pointers? If any of those is true, Extrae will not be able to capture MPI calls as it only intercepts MPI_* functions.

**Comment by vineetsoni** [comment]
The intermediate library does not use the PMPI interface. It makes standard MPI_* function calls.
Would you suggest that if MPI_* calls are replaced with PMPI_*, Extrae would be able to intercept them?

**Comment by gllort** [comment]
No, using PMPI_* calls directly will definitely not work, you need to stick to the MPI_* calls for Extrae to be able to intercept them.
